Enable search param - imgSize in lowercase entry during programmatic usage

SiewLinYap opened the pull request:
To allow `imgSize` to be specified in lowercase just as other search params. It is not clear that this particular param needs to  be in all uppercase. Readme also shows the input for this param is in lowercase. 

User will not notice this until hits error in terminal with message
```
TypeError: Parameter "imgSize" value "medium" is not an allowed value in "['imgSizeUndefined', 'HUGE', 'ICON', 'LARGE', 'MEDIUM', 'SMALL', 'XLARGE', 'XXLARGE']"
```


Notice for CLI usage [ `cli.py` line 60-61 ]
```
    if imagesize:
        imagesize = imagesize.upper()
```
there’s a special handling of this param to Uppercase. It will be good to allow the same special handling done internally in `fetch_resize_save.py` as well if user choose the programmatic usage

arrrlo commented:
Hi @SiewLinYap 

Thanks for submitting.

If this parameter is casted to uppers with every request, how do we handle `imgSizeUndefined` then?

SiewLinYap commented:
Hi @arrrlo  
Thanks for this package development as well. It saves us time to get the images ourselves. 

For the question on handling `imgSizeUndefined`, kindly see my reply below:

Under class `GoogleCustomSearch `, 
```
        self._search_params_keys = {
            'q': None,
            'searchType': 'image',
            'num': 1,
            'start': 1,
            'imgType': None,
            'imgSize': None,
            'fileType': None,
            'safe': 'off',
            'imgDominantColor': None,
            'imgColorType': None,
            'rights': None
        }
```

`imgSize` in `self._search_params_keys` is initialised as `None`.

In internal method of the same class, `_search_params` is proposed to change as below:
```
    def _search_params(self, params):
        """Received a dict of params and merges
        it with default params dict
        :param params: dict
        :return: dict
        """

        search_params = {}

        for key, value in self._search_params_keys.items():
            params_value = params.get(key)

            if params_value:
                if key == "imgSize":
                    params_value = params_value.upper()
                # take user defined param value if defined
                search_params[key] = params_value
            elif value:
                # take default param value if defined
                search_params[key] = value

        return search_params
```

For scenario when imgSize is not even included in user’s search params ( == `imgSize` not mentioned at all ), the `params_value` will return `None`, and since the next check is `elif value` (which is initialised as `None` as well ), then this parameter `imgSize` will just pass and the final `search_params` dict will not have this para included and search will proceed without taking this para into consideration

For scenario if user don’t know what to put for this `imgSize` and somehow put as `imgSize: ‘’`(empty), this will also cause both `if` and `elif` get passed and eventually have the same results which final `search_params` will not have this para included.

So in either cases above, the search and download can still proceed.

Since in the `README`,  there’s no mentioned about `imgSizeUndefined` as an option for `imgSize` when user doesn’t want to have any specific filter for this parameter, this particular option value is unlikely to fail as `imgSizeUndefined` is not even visible to user unless they fail/hit error in terminal. User will most likely either just skip this param, put as `None` or ``, rather than specify it to `imgSizeUndefined` when they don’t want to consider this para and don't even know it shall be input as such ..

Even if user refer to the [documentation](https://developers.google.com/custom-search/v1/reference/rest/v1/ImgSize), the actual undefined value is also different from what is mentioned as `imgSizeUndefined` when `TypeError` occured. It looks more like either `IMG_SIZE_UNDEFINED` or just `UNDEFINED`… I’m not sure where this `imgSizeUndefined` is stored/first defined, if really want to consider this specific option, perhaps an update to `README` is needed also along with the inclusion check of `imgSizeUndefined` in internal method `_search_params`…

I may have misunderstood the original intention/usage of having `imgSizeUndefined` in the allowable list for `imgSize`, in such case, kindly clarify. Thanks.

arrrlo commented:
Hi @SiewLinYap 

Thank you for this extensive and detailed explanation.

And yes, currently, the lib fails in doing the right job regarding what you described.
However, let's do this right.
Documentation should include `imgSizeUndefined` and the code should handle it right (combination of lowers and uppers) along with other values (all uppers).

Can you make changes to comply with this?

SiewLinYap commented:
Hi @arrrlo 

I’ve updated the method `def _search_params(self, params)` to have conditional internal check for `imgSizeUndefined` while transforming other `imgSize` options to upper-case to align with the google custom search allowable list for this parameter.

Also, updated the `README` to include scenario of search param value xxx is undefined. Think it may not be sufficient to just update the undefined option for `imgSize` while other params  actually also have this option. Therefore, take this PR to include the readme update for those params as well even though in this PR, only script related to`imgSize` is affected due to its special case which values are somehow pre-defined in uppercase form in the incoming google custom search api call.

Key changes made to `README` as below :
- Add comments to guide user which search param is not multiselect-enabled and thus shall only input one option at a time [ To avoid users input something like `large|medium|small` as the value for `imgSize` or other search param]
- Include xxxUndefined for affected search keys, not limited to just `imgSize`
- Update the options for `safe` to include `active`. According to the [documentaion](https://developers.google.com/custom-search/v1/reference/rest/v1/Safe), `high` and `medium` is deprecated and they are equivalent to `active`. But since both `high` and `medium` still not fully deprecated yet till date, so I retain both along with the newly added `active`
- Remove `news` as option for `imgType` as this is no longer valid. Instead, add new available options `stock` and `animated`

Please check and review if changes above aligned to what you expect as well.

arrrlo commented:
Hi @SiewLinYap 

Great!
Let's merge this PR.

Thank you very much for your efforts.
